src/theory/bv/theory_bv.h

```diff
@@ -115,177 +99,141 @@ class TheoryBV : public Theory {
 
   void presolve() override;
 
+  /** Called by abstraction preprocessing pass. */
   bool applyAbstraction(const std::vector<Node>& assertions,
                         std::vector<Node>& new_assertions);
 
  private:
-  class Statistics
-  {
-   public:
-    AverageStat d_avgConflictSize;
-    IntStat     d_solveSubstitutions;
-    TimerStat   d_solveTimer;
-    IntStat     d_numCallsToCheckFullEffort;
-    IntStat     d_numCallsToCheckStandardEffort;
-    TimerStat   d_weightComputationTimer;
-    IntStat     d_numMultSlice;
-    Statistics();
-    ~Statistics();
-  };
-
-  Statistics d_statistics;
-
-  void spendResource(ResourceManager::Resource r);
+  void notifySharedTerm(TNode t) override;
 
   /**
-   * Return the uninterpreted function symbol corresponding to division-by-zero
-   * for this particular bit-width
+   * Return the UF symbol corresponding to division-by-zero for this particular
+   * bit-width.
    * @param k should be UREM or UDIV
-   * @param width
-   *
-   * @return
+   * @param width bit-width
    */
-  Node getBVDivByZero(Kind k, unsigned width);
+  Node getUFDivByZero(Kind k, unsigned width);
+
+  /** Internal BV solver. */
+  std::unique_ptr<TheoryBVSolver> d_internal;
 
-  typedef std::unordered_set<TNode, TNodeHashFunction> TNodeSet;
-  typedef std::unordered_set<Node, NodeHashFunction> NodeSet;
-  NodeSet d_staticLearnCache;
+  /** Extended theory module, for context-dependent simplification. */
+  std::unique_ptr<ExtTheory> d_extTheory;
 
   /**
    * Maps from bit-vector width to division-by-zero uninterpreted
    * function symbols.
    */
-  std::unordered_map<unsigned, Node> d_BVDivByZero;
-  std::unordered_map<unsigned, Node> d_BVRemByZero;
+  std::unordered_map<unsigned, Node> d_ufDivByZero;
+  std::unordered_map<unsigned, Node> d_ufRemByZero;
 
-  typedef std::unordered_map<Node, Node, NodeHashFunction>  NodeToNode;
+  /** The theory rewriter for this theory. */
+  TheoryBVRewriter d_rewriter;
 
-  context::CDO<bool> d_lemmasAdded;
+  /** A (default) theory state object */
+  TheoryState d_state;
 
-  // Are we in conflict?
-  context::CDO<bool> d_conflict;
+}; /* class TheoryBV */
 
-  // Invalidate the model cache if check was called
-  context::CDO<bool> d_invalidateModelCache;
+class TheoryBVSolver
```

Comment:
I think this interface class might be overkill.  I think the end goal is to replace and delete the old solver, in which case we won't have multiple instances of this class.

Concretely my suggestion would be to have TheoryBv maintain a TheoryBvLazy class that is used temporarily while we migrate to the new solver.  Then this would allow the new implementation either to live directly within TheoryBV, or  otherwise the new solver would not have to be restricted to follow this interface (which is the job of TheoryBV anyways).